### DISCUSSION TITLE
ld: warning: ignoring duplicate libraries: '-lemutls_w', '-lgcc'

### DISCUSSION THREAD

**Discussion by canecaster** [Homebrew/discussions#4794] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.12
ORIGIN: https://github.com/Homebrew/brew
HEAD: aa2a77b1f1ac1539e69e186bad1d569100448987
Last commit: 2 days ago
Core tap JSON: 18 Sep 21:10 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: dodeca-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.5.2-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: 14.3.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

MacOS System Settings prompted me to update the Command Line Tools today, which I did.  At that point I started getting the warning "ld: warning: ignoring duplicate libraries: '-lemutls_w', '-lgcc'" when I would run gfortran.  I executed brew update and brew upgrade but the warning message persists.  Compiled Fortran code seems functional when recompiled but unsure whether to worry about this warning or how to resolve.  Guidance would be very much appreciated.

James

**Comment by fxcoudert** [comment, marked as answer]
Yes, this is an incompatibility of current gcc with the Xcode 15 and/or CLT 15. The warning is not a problem, and we will ship a fixed version in the future.

**Reply by canecaster** [reply]
Thank you!

**Reply by subhodynamics** [reply]
Since this version of CLT got a problem with current gcc, and will require a new update, meanwhile what would you suggest us to do, to continue our works.
Thanks.

**Reply by fxcoudert** [reply]
At this stage, best option is: do not update Xcode / CLT.
However, ss I said, the warning is harmless.

**Reply by subhodynamics** [reply]
i am unable to compile my cpp files using gcc (homebrew) since I updated my CLT, but then i tried downgrading it, it still is not working in my case.

**Reply by osalbahr** [reply]
> i am unable to compile my cpp files using gcc (homebrew) since I updated my CLT, but then i tried downgrading it, it still is not working in my case.

~~Just to understand correctly, are you unable to use Homebrew's gcc even with Xcode 14? Could you provide the output of `brew doctor` and `brew config`?~~

Edit: nevermind, I saw you already did paste a screenshot below.

**Reply by HeavenlyBard** [reply]
只需使用“clang helloworld.o -o helloworld”，请忘记-lSystem。

**Comment by black-ole** [comment]
I have the same issue. My output of brew config is same as yours and when I am compiling my c++ file using gcc. It gives errors. I hope this fixes soon, if its the issue of Xcode CLT 15
```
blackole@Divyanss-MacBook-Air ~ % g++-13 test.cpp
ld: warning: ignoring duplicate libraries: '-lgcc'
0  0x100363648  __assert_rtn + 72
1  0x100297fac  ld::AtomPlacement::findAtom(unsigned char, unsigned long long, ld::AtomPlacement::AtomLoc const*&, long long&) const + 1204
2  0x1002ad924  ld::InputFiles::SliceParser::parseObjectFile(mach_o::Header const*) const + 15164
3  0x1002bae30  ld::InputFiles::parseAllFiles(void (ld::AtomFile const*) block_pointer)::$_7::operator()(unsigned long, ld::FileInfo const&) const + 420
4  0x18f1b4440  _dispatch_client_callout2 + 20
5  0x18f1c9544  _dispatch_apply_invoke_and_wait + 224
6  0x18f1c884c  _dispatch_apply_with_attr_f + 1180
7  0x18f1c8a38  dispatch_apply + 96
8  0x1003353b8  ld::AtomFileConsolidator::parseFiles(bool) + 292
9  0x1002d6170  main + 9048
ld: Assertion failed: (resultIndex < sectData.atoms.size()), function findAtom, file Relocations.cpp, line 1336.
collect2: error: ld returned 1 exit status
```

**Reply by fxcoudert** [reply]
This is a different issue, but it is also caused by Xcode / CLT 15. It is a bug in Apple's linker, which was reported to them in August. So far, they have not released a fix.

**Reply by black-ole** [reply]
Ya, I rolled back to Xcode CLT 14, now issue is resolved.

**Reply by dask-58** [reply]
@black-ole Dude plz elaborate what you did I am facing the same issue.

**Reply by prem3402** [reply]
Uninstall Xcode CLT 15 using " sudo rm -rf /Library/Developer/Xcode " and download Xcode CLT 14 from " https://developer.apple.com/download/all/ " this website and extract the package.

**Reply by dask-58** [reply]
@Mrunrighteous So is it because of an update my Mac got recently? It was for clt 
Thanks for your help

**Reply by black-ole** [reply]
Yes, its because of update. Specifically Xcode CLT

**Reply by dask-58** [reply]
@Mrunrighteous and @black-ole thank you so much guys. Now my problem is fixed. I was so worried like what the heck is this 😅

**Reply by subhodynamics** [reply]
@dask-58 can you confirm exactly which CLT version you downgraded to, i downgraded to 14.2, and still there is a compilation error

**Reply by dask-58** [reply]
@subhodynamics I installed clt 14.3.1

**Reply by subhodynamics** [reply]
@dask-58 I tried installing 14.3 but its not working in my case. it says  
<img width="607" alt="Screenshot 2023-09-23 at 11 59 23 AM" src="https://github.com/Homebrew/discussions/assets/75271901/410d6784-9fe2-425f-9c4b-2a168ed6e756">


This is my config
<img width="668" alt="Screenshot 2023-09-23 at 11 36 59 AM" src="https://github.com/Homebrew/discussions/assets/75271901/cde32ed8-7ff6-429b-9597-c49a71450d99">

**Reply by dask-58** [reply]
Sorry @subhodynamics but I don't really know what that means just make sure you uninstalled clt15 properly and installed the correct version of CLT 14.3.1

**Reply by saifsmailbox98** [reply]
I think there's no need to remove the existing files. The installer does it by default.

**Reply by AGaillardTSE** [reply]
@subhodynamics, it is an issue with using std libraries. If you compile using g++ instead of gcc it should be fixed.

**Reply by dask-58** [reply]
@AGaillardTSE so does compiling with g++ also work in case of Xcode CLT 15?

**Reply by AGaillardTSE** [reply]
@dask-58 , for the problem reported above by  black-ole, it does nothing. But for the specific problem of @subhodynamics with "symbol(s) not found for architecture arm64" it does.

**Reply by dask-58** [reply]
I had the same issue as @black-ole 's which was fixed when I downgraded clt. Anyway thanks 🙏

**Reply by wadidf** [reply]
Add in the linker flags the following : 
-Wl,-ld_classic

**Reply by saenkoyd** [reply]
* linker flags in Xcode

Works for Xcode 15, not 14. Thank you!

Is it safe – don't know :D

**Reply by virksukh** [reply]
@wadidf , can you please tell me how can in the linker the flags that you have mentioned about . Please help.. I am facing the same issue, whenI downgraded to Xcode 14.3.1 CLT, it show  "Command Line Tools can't be installed on this disk. The version of macOS is too new" error. I want to try out your solution but doesn't know how can I implement your solution in practicality.

**Reply by shubhampatel27** [reply]
hi, you find out the solution?? actually i am also trying to troubleshoot the problem and i am able to add linker flag in Xcode .
can you please guide how you find out the solution??

**Reply by Goel-Akash** [reply]
How do I add the linker flags? I can't downgrade as my Mac version is too new.

**Reply by dask-58** [reply]
compile your code using 
g++ -Wl,-ld_classic -o executable_name file_name.cpp

**Reply by Goel-Akash** [reply]
Any other way this problem could be solved?  Perhaps using something other than VS code

**Reply by wadidf** [reply]
Adding linked flags depends on how you are coding (tools). 
Examples :

**-  using gcc-13 (you can replace with g++-13 and with respective version ** 

gcc-13 -o output_file_name input_file_name.c -Wl,-ld_classic 

**- using XCode** 
in your project properties go to 'Build Settings' -> 'All' -> 'Other Linker Flags' -> '+' 
and write (or paste) the flags **-Wl,-ld_classic**

The answer to 'is it safe or not' : Yes, it's a feature that Apple left in the linker to allow smooth transition to newer versions.

The meaning of the flag is : please use the old and 'used for some time' linker instead of the newer (and problematic) linker

**Reply by spinicist** [reply]
@wadidf Thank you SO MUCH. I have been having random crashes (including illegal instructions!) since upgrading XCode. Adding `-Wl,-ld_classic` seems to have fixed it. What the heck did Apple do?

**Reply by spinicist** [reply]
@fxcoudert Did Apple confirm this in public? And if so, where? I have reported my issue in Feedback Assistant but had no response.

**Reply by wadidf** [reply]
@spinicist https://developer.apple.com/documentation/xcode-release-notes/xcode-15-release-notes
[](https://developer.apple.com/documentation/xcode-release-notes/xcode-15-release-notes)

**Reply by spinicist** [reply]
@wadidf Thanks again

**Reply by the13fools** [reply]
try reinstalling gcc as indicated by @Goel-Akash below, that helped me!

**Comment by Harshal5** [comment]
> @subhodynamics I installed clt 14.3.1

@dask-58 I am not able to install CLT 14.3.1, as while "selecting the disk" stage of installation I get the "Command Line Tools can't be installed on this disk. The version of macOS is too new" error.
Could you let me know how did you install the 14.3.1 version?

**Reply by AkshatAggarwal14** [reply]
You don't need to downgrade your CLT version. [This solution](https://github.com/orgs/Homebrew/discussions/4794#discussioncomment-7117550) works!

**Reply by dask-58** [reply]
Thank you so much!. But is it safe? one of the previous replies doesn't guarantee that

**Reply by fxcoudert** [reply]
What do you mean by safe? It uses the older version of the linker, which was used until Xcode 15.

**Reply by wadidf** [reply]
Please read my answer to @Goel-Akash

**Reply by Goel-Akash** [reply]
Just putting it out there... 
My problem did get solved as i did
brew install gcc
I did some updates and my problem got solved.
All my cpp codes are working in vscode without any problem or warning

**Reply by shubhampatel27** [reply]
hi Goel-Akash ,
i am still trying to find out the solution can you please tell me which kind of updates you did ,it will really helpful.

**Reply by Goel-Akash** [reply]
I just wrote 
brew install gcc 
In terminal and everything fixed on its own

**Reply by shubhampatel27** [reply]
finally its working from last 2 days i am trying to resolve this..
Thank you so much really appreciate

**Comment by virksukh** [comment]
Hi Shubham,
I just reinstalled MacOS Sonoma, actually I had installed with Sonoma so
but there were problems like the one above so I discussed with MacOS
support and they advised me to reisntall everything.. I did and thankfully
the problems are solved (for now, fingers crossed)



On Tue, 17 Oct 2023 at 09:45, Shubham patel ***@***.***>
wrote:

> hi, you find out the solution?? actually i am also trying to troubleshoot
> the problem and i am able to add linker flag in Xcode .
> can you please guide how you find out the solution??
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/Homebrew/discussions/4794#discussioncomment-7299566>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABW7UIYPX7DMA34LH7BHVBLX7YA4TAVCNFSM6AAAAAA45KSQ7GVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TEOJZGU3DM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Sukhdeep

**Reply by shubhampatel27** [reply]
Thank you so much, I really appreciate your effort and value your time.
i will also try to do this hope it will work.
Thanks.

**Comment by rdk0215** [comment]
For those who are still facing issues  try this :
1. uninstall gcc using : brew uninstall gcc 
2. reinstall gcc using : brew install gcc 
 Worked for me :)

**Reply by canecaster** [reply]
Just to clarify, because my original issue was a duplicate libraries warning.  Then later there was discussion about actual compiler errors.  Wondering which issue reinstalling gcc fixed for you.

James

**Reply by mohitsom29** [reply]
I did what you have said. But still, it's not working for me in my sublime text editor. I tried to run my C++ code in sublime text and it shows nothing and the code gets stuck. No error is coming simply the code gets stuck.

**Reply by owainkenwayucl** [reply]
I think he means "fix the duplicate -lgcc warning".

As a note, if like me you have lots of things that depend on gcc, you can do `brew reinstall gcc` which works just as well and fixed the duplicate -lgcc warning for me.

**Reply by canecaster** [reply]
Thank you!  I did a "brew reinstall gcc" and the duplicate libraries warning indeed disappeared.
